✨ feat: split ITS PCHs into status and clusteradmin (#3024)

rishi-jat opened the pull request:
fixes #3024

rishi-jat commented:
@MikeSpreitzer

MikeSpreitzer commented on core-chart/templates/postcreatehooks/install-status-addon.yaml:
```diff
@@ -0,0 +1,69 @@
+                secretName: "{{"{{.ITSSecretName}}"}}"
+          restartPolicy: Never
+      backoffLimit: 1
+{{- end }} 
```
Is there any way that I could get you to have a newline at the end of a file, so that GitHub does not display the red circle-dash alert?
<img width="210" height="78" alt="image" src="https://github.com/user-attachments/assets/6f631dc3-fa7e-4677-a90e-525c6cb9eade" />

MikeSpreitzer commented on scripts/kubestellar-snapshot.sh:
```diff
@@ -420,9 +420,15 @@ for i in "${!cps[@]}" ; do # for all control planes in context ${context}
+      cp_pchs=($(kubectl --context $helm_context get controlplane $name -o jsonpath='{.spec.postCreateHook}'))
+    fi
+    # Print control plane info
+    pch_list="${cp_pchs[*]}"
```
Why introduce `pch_list`? Why not just use `cp_pchs` instead in the next command?

MikeSpreitzer commented:
/ok-to-test

MikeSpreitzer commented on core-chart/templates/postcreatehooks/its-clusteradmin.yaml:
Why is this file here?

MikeSpreitzer commented on core-chart/templates/postcreatehooks/its-status.yaml:
Why is this file here?

MikeSpreitzer commented:
@rishi-jat: You should be testing your contributions.

francostellari commented:
@rishi-jat your failure https://github.com/kubestellar/kubestellar/actions/runs/16496875610/job/46644656573?pr=3121#step:7:227 is due to the wrong syntax in listing the hooks.

See the WDS CP definition as ana example: https://github.com/kubestellar/kubestellar/blob/main/core-chart/templates/controlplanes/wds.yaml

rishi-jat commented:
/retest

rishi-jat commented:
Please have a look and let me know if any further improvements are needed.

MikeSpreitzer commented on docs/content/direct/get-started.md:
````diff
@@ -141,8 +141,10 @@ kflex ctx --overwrite-existing-context its1
 ```shell
-kubectl --context kind-kubeflex wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-with-clusteradm}=true' --timeout 90s
-kubectl --context kind-kubeflex wait -n its1-system job.batch/its-with-clusteradm --for condition=Complete --timeout 150s
+kubectl --context kind-kubeflex wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.install-status-addon}=true' --timeout 90s
+kubectl --context kind-kubeflex wait -n its1-system job.batch/install-status-addon --for condition=Complete --timeout 150s
+kubectl --context kind-kubeflex wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-hub-init}=true' --timeout 90s
+kubectl --context kind-kubeflex wait -n its1-system job.batch/its-hub-init --for condition=Complete --timeout 150s
````
@francostellari , @Rupam-It: it is now sufficient to wait for the `ControlPlane` to be "ready", because that encompases both PCHs having their `Job` objects created and finished running, right?

francostellari replied:
@MikeSpreitzer , not yet.
I have opened issue #3145

francostellari commented:
@rishi-jat please delete the `_helpers.tpl` file and move the 2 sections of code inside the 2 PCHs.
Thank you.

rishi-jat commented:
@MikeSpreitzer @francostellari I've completed the inlining of both PCH templates and removed `_helpers.tpl` as requested. Please take another look.

MikeSpreitzer commented:
The synchronization problem could be fixed by adding to the Job. that installs the status add-on an initContainer that does `kubectl wait` for the CRD for `ManagedClusterSetBinding` to be "established".

MikeSpreitzer commented:
No, I was wrong, the current ITS init Job does not use two initContainers, it uses two plain containers. That would explain some flakiness.

MikeSpreitzer commented on core-chart/templates/postcreatehooks/install-status-addon.yaml:
```diff
@@ -0,0 +1,107 @@
+              - |
+                echo "Waiting for OCM CRDs to be established..."
+                # Wait for core OCM CRDs that clusteradm creates
+                kubectl wait --for=condition=Established crd/managedclusters.cluster.open-cluster-management.io --timeout=300s
+                kubectl wait --for=condition=Established crd/managedclustersets.cluster.open-cluster-management.io --timeout=300s
+                kubectl wait --for=condition=Established crd/managedclustersetbindings.cluster.open-cluster-management.io --timeout=300s
+                kubectl wait --for=condition=Established crd/placements.cluster.open-cluster-management.io --timeout=300s
+                kubectl wait --for=condition=Established crd/placementdecisions.cluster.open-cluster-management.io --timeout=300s
```
Do we really need to wait for all of these?
Are all of these used while installing the status add-on?

MikeSpreitzer commented on core-chart/templates/postcreatehooks/install-status-addon.yaml:
```diff
@@ -0,0 +1,107 @@
+              mountPath: "/etc/kube"
+              readOnly: true
+          containers:
+          - name: "{{"{{.HookName}}-statusaddon"}}"
```
FYI, "-statusaddon" is redundant with the hook name.

MikeSpreitzer commented:
@francostellari: yes, I wrote something wrong. In https://github.com/kubestellar/kubestellar/pull/3121#issuecomment-3134518927 I pointed to another copy of the correction.

To restate, there were and remain two synchronization problems. This PR should solve both of them, we are tired of the flakes.

- Before using it, ensure that the in-cluster data member exists in the Secret that holds the kubeconfigs for accessing the ITS.
- Ensure that the OCM hub initialization has finished before starting to install the OCM Status Add-On.

MikeSpreitzer commented on core-chart/templates/postcreatehooks/its-hub-init.yaml:
```diff
@@ -0,0 +1,50 @@
+              - -c
+              - |
+                echo "Waiting for config-incluster in Secret {{.ITSSecretName}}..."
+                until kubectl get secret "{{.ITSSecretName}}" -n {{.ControlPlaneName}}-system -o jsonpath="{.data.config-incluster}" | grep -q .; do
```
This `kubectl` command runs with `$KUBECONFIG` referring to the thing it is intended to wait for: the in-cluster kubeconfig data member of the Secret holding kubeconfigs for accessing the ITS. This is bad because it is circular, worse because it is looking in the wrong place. The Secret holding the kubeconfigs for accessing the ITS is in the KubeFlex hosting cluster.

MikeSpreitzer replied:
There is good news and bad news for you in https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod:

> When a volume contains data from a Secret, and that Secret is updated, Kubernetes tracks this and updates the data in the volume, using an eventually-consistent approach.

What this initContainer needs to accomplish is to ensure that when the regular container tries to use `/etc/kube/{{ .ITSkubeconfig }}`, that file exists. Because the kubelet's update of the volume is only eventually consistent with the Secret, waiting for the data member to exist in the Secret is not enough.

The good news is that it is enough to wait for the file to appear.

rishi-jat commented:
I’ve implemented all your review suggestions for PR :
	•	Fixed the synchronization gap by replacing the circular kubectl get secret logic with a proper check for           kubeconfig file existence.
	•	Reduced CRD waiting to only the essential ones: placements and placementdecisions.
	•	Added an initContainer to wait for the kubeconfig file before proceeding.
	•	Removed the redundant -statusaddon suffix from the container name.
	•	Cleaned up unnecessary whitespace.

MikeSpreitzer commented:
> Reduced CRD waiting to only the essential ones: placements and placementdecisions

Those are not the two I saw in the log I looked at recently. There it was placement and (something)managed(something).

MikeSpreitzer commented on core-chart/templates/postcreatehooks/install-status-addon.yaml:
```diff
@@ -0,0 +1,123 @@
+                max_attempts=60
+                attempt=0
+                while [ $attempt -lt $max_attempts ]; do
+                  if kubectl get crd placements.cluster.open-cluster-management.io >/dev/null 2>&1; then
```
Is it enough to only wait for one of them to be created? Are we guaranteed that this one is created last?

MikeSpreitzer commented on core-chart/templates/postcreatehooks/install-status-addon.yaml:
```diff
@@ -0,0 +1,123 @@
+        spec:
+          initContainers:
+          - name: wait-for-kubeconfig
+            image: quay.io/kubestellar/kubectl:{{.Values.KUBECTL_VERSION}}
```
Could we use `busybox` instead of our kubectl image? We want to work towards removing `bash` from our kubectl image.

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -197,4 +201,4 @@ if ! expect-cmd-output 'kubectl --context its1 get managedclusters -l location-g
     echo "Failed to see two clusters."
     exit 1
-fi
+fi
```
Please do not delete the linebreak at the end of the file. That deletion makes GitHub display an alarming character here:

<img width="204" height="110" alt="image" src="https://github.com/user-attachments/assets/ff7cc36f-7c94-477f-ac23-aa30c0ea7e54" />

MikeSpreitzer reviewed (approved):
More work can be done in follow-on PRs. This is good enough to merge. It looks like it will help with #2201 (#2717).

/lgtm
/approve
Thanks!

MikeSpreitzer commented on core-chart/templates/postcreatehooks/install-status-addon.yaml:
```diff
@@ -0,0 +1,123 @@
+              mountPath: "/etc/kube"
+              readOnly: true
+          containers:
+          - name: "{{"{{.HookName}}"}}"
```
This container exists within the context of a Pod that is specific to this PCH. In my opinion, the name of this hook does not need to be repeated here.

MikeSpreitzer commented on docs/content/direct/get-started.md:
````diff
@@ -141,8 +141,10 @@ kflex ctx --overwrite-existing-context its1
 ```shell
-kubectl --context kind-kubeflex wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-with-clusteradm}=true' --timeout 90s
-kubectl --context kind-kubeflex wait -n its1-system job.batch/its-with-clusteradm --for condition=Complete --timeout 150s
+kubectl --context kind-kubeflex wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.install-status-addon}=true' --timeout 90s
+kubectl --context kind-kubeflex wait -n its1-system job.batch/install-status-addon --for condition=Complete --timeout 150s
+kubectl --context kind-kubeflex wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-hub-init}=true' --timeout 90s
+kubectl --context kind-kubeflex wait -n its1-system job.batch/its-hub-init --for condition=Complete --timeout 150s
````
OOPS! I should not have approved this change. It is not right when the latest release of the Helm chart is being used --- that has not changed yet.

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -163,8 +165,10 @@ function add_wec() {
 
 "${SRC_DIR}/../../../scripts/check_pre_req.sh" --assert --verbose ocm
 
-kubectl --context $HOSTING_CONTEXT wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-with-clusteradm}=true' --timeout 200s
-kubectl --context $HOSTING_CONTEXT wait -n its1-system job.batch/its-with-clusteradm --for condition=Complete --timeout 400s
+kubectl --context $HOSTING_CONTEXT wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.install-status-addon}=true' --timeout 200s
+kubectl --context $HOSTING_CONTEXT wait -n its1-system job.batch/install-status-addon --for condition=Complete --timeout 400s
+kubectl --context $HOSTING_CONTEXT wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-hub-init}=true' --timeout 200s
+kubectl --context $HOSTING_CONTEXT wait -n its1-system job.batch/its-hub-init --for condition=Complete --timeout 400s
```
BTW, this is a (functionally equivalent but textually different) copy of an earlier block of code.
I do not see why it should be here.
I introduced the duplication in #2546, perhaps unaware that I was making a duplication.

rishi-jat commented:
@MikeSpreitzer Thanks for the feedback. I’ll address all of these switching to busybox, waiting for the correct CRDs, and cleaning up formatting  in a follow-up PR soon.

MikeSpreitzer commented on core-chart/templates/postcreatehooks/its-hub-init.yaml:
```diff
@@ -0,0 +1,56 @@
+              - sh
+              - -c
+              - |
+                echo "Waiting for kubeconfig file /etc/kube/{{.ITSkubeconfig}} to exist and be non-empty..."
```
That is due to a quoting problem. To debug it, I changed the source to the following.

```yaml
            command:
              - sh
              - -c
              - |
                echo "Form 1: /etc/kube/{{.ITSkubeconfig}}"
                echo Form 2:  /etc/kube/{{.ITSkubeconfig}}
                echo Form 3: "/etc/kube/{{.ITSkubeconfig}}"
                echo Form 4:  {{"/etc/kube/{{.ITSkubeconfig}}"}}
                echo Form 5: "{{"/etc/kube/{{.ITSkubeconfig}}"}}"
                echo "Waiting for kubeconfig file '$KUBECONFIG' to exist and be non-empty..."
                until [ -s "$KUBECONFIG" ]; do
                  echo "Kubeconfig file not ready yet, waiting..."
                  sleep 2
                done
                echo "Kubeconfig file is ready"
            env:
            - name: KUBECONFIG
              value: "{{"/etc/kube/{{.ITSkubeconfig}}"}}"
```

After the Pod was constructed and run, a `kubectl get -o yaml` of it showed me the following.

```yaml
  - command:
    - sh
    - -c
    - |
      echo "Form 1: /etc/kube/"
      echo Form 2:  /etc/kube/
      echo Form 3: "/etc/kube/"
      echo Form 4:  /etc/kube/config-incluster
      echo Form 5: "/etc/kube/config-incluster"
      echo "Waiting for kubeconfig file '$KUBECONFIG' to exist and be non-empty..."
      until [ -s "$KUBECONFIG" ]; do
        echo "Kubeconfig file not ready yet, waiting..."
        sleep 2
      done
      echo "Kubeconfig file is ready"
    env:
    - name: KUBECONFIG
      value: /etc/kube/config-incluster
```
